File: renu_customization/renu_customization/page/collection_report/collection_report.py
import frappe
from frappe import _
from frappe.utils import flt, nowdate, now_datetime
import openpyxl
from openpyxl.styles import Alignment, Font, PatternFill, Border, Side
from openpyxl.utils import get_column_letter
from io import BytesIO
import base64

@frappe.whitelist()
def export_to_pdf(html):
    pdf_content = frappe.utils.pdf.get_pdf(html, {"orientation": "Landscape"})
    frappe.response.filename = f"Collection_Report_{nowdate()}.pdf"
    frappe.response.filecontent = pdf_content
    frappe.response.type = "download"

def prepare_filters(filters):
    if not filters:
        filters = {}
    elif isinstance(filters, str):
        filters = frappe.parse_json(filters)
    
    # Handle DateRange from JS
    if filters.get("date_range"):
        dr = filters.get("date_range")
        if isinstance(dr, list) and len(dr) == 2:
            filters["from_date"] = dr[0]
            filters["to_date"] = dr[1]

    # Handle Fiscal Year
    if filters.get("fiscal_year") and (not filters.get("from_date") or not filters.get("to_date")):
        fy = frappe.get_doc("Fiscal Year", filters.get("fiscal_year"))
        if fy:
            filters["from_date"] = filters.get("from_date") or fy.year_start_date
            filters["to_date"] = filters.get("to_date") or fy.year_end_date
    
    return frappe._dict(filters)

@frappe.whitelist()
def get_dashboard_data(filters=None):
    filters = prepare_filters(filters)

    conditions = [
        "pe.docstatus = 1",
        "pe.payment_type = 'Receive'",
        "pe.party_type = 'Customer'",
        "per.reference_doctype = 'Sales Invoice'"
    ]
    
    if filters.get("customer"):
        conditions.append(f"pe.party = {frappe.db.escape(filters.get('customer'))}")
    
    if filters.get("from_date"):
        conditions.append(f"pe.posting_date >= {frappe.db.escape(filters.get('from_date'))}")
    
    if filters.get("to_date"):
        conditions.append(f"pe.posting_date <= {frappe.db.escape(filters.get('to_date'))}")
 
    if filters.get("dom_exp"):
        if filters.get("dom_exp") == "Domestic":
            conditions.append("si.is_domestic = 1")
        elif filters.get("dom_exp") == "Export":
            conditions.append("si.is_export = 1")
 
    where_clause = " AND ".join(conditions)
    
    query = f"""
        SELECT 
            pe.name as payment_entry,
            pe.posting_date,
            pe.party as customer,
            per.reference_name as name,
            per.base_allocated_amount,
            si.is_export, 
            si.is_domestic,
            si.status,
            sii.item_code,
            sii.item_name,
            sii.base_amount as item_amount,
            si.base_net_total
        FROM `tabPayment Entry` pe
        JOIN `tabPayment Entry Reference` per ON per.parent = pe.name
        JOIN `tabSales Invoice` si ON si.name = per.reference_name
        JOIN `tabSales Invoice Item` sii ON sii.parent = si.name
        WHERE {where_clause}
        ORDER BY pe.posting_date DESC, pe.name DESC
    """
    
    data = frappe.db.sql(query, as_dict=True)
    
    # Fetch Sales Person from Sales Team child table
    inv_names = list(set([d.name for d in data]))
    sales_map = {}
    if inv_names:
        sales_team = frappe.get_all("Sales Team", 
            filters={"parent": ("in", inv_names), "parenttype": "Sales Invoice"},
            fields=["parent", "sales_person"]
        )
        for st in sales_team:
            if st.parent not in sales_map:
                sales_map[st.parent] = []
            sales_map[st.parent].append(st.sales_person)
    
    # Process data and apply sales person filter
    final_data = []
    sp_filter = filters.get("sales_person")
    
    # Track unique allocations for KPI calculations to avoid double counting due to item join
    unique_allocations = {}
    
    for d in data:
        d.sales_person = ", ".join(sales_map.get(d.name, []))
        
        if sp_filter and sp_filter not in (d.sales_person or ""):
            continue
        
        # Format Item
        d.item = f"{d.item_code} {d.item_name}" if d.item_code else (d.item_name or "")
        
        # Calculate item's share of the allocated payment (Pro-rata allocation)
        if flt(d.base_net_total) > 0:
            d.allocated_amount = (flt(d.item_amount) / flt(d.base_net_total)) * flt(d.base_allocated_amount)
        else:
            d.allocated_amount = flt(d.item_amount)
            
        final_data.append(d)
        
        # For KPIs, we sum unique payment-to-invoice allocations
        alloc_key = f"{d.payment_entry}-{d.name}"
        if alloc_key not in unique_allocations:
            unique_allocations[alloc_key] = {
                "amount": flt(d.base_allocated_amount),
                "is_export": d.is_export,
                "is_domestic": d.is_domestic
            }
 
    # Calculate KPIs from unique allocations - round each value to 4 decimal places in M
    total_collection = sum(round(v["amount"] / 1000000, 4) for v in unique_allocations.values()) * 1000000
    export_collection = sum(round(v["amount"] / 1000000, 4) for v in unique_allocations.values() if v["is_export"]) * 1000000
    domestic_collection = sum(round(v["amount"] / 1000000, 4) for v in unique_allocations.values() if not v["is_export"]) * 1000000
    
    summary = [
        {"label": _("TOTAL Collection"), "value": total_collection, "indicator": "Blue"},
        {"label": _("Export Collection"), "value": export_collection, "indicator": "Green"},
        {"label": _("Domestic Collection"), "value": domestic_collection, "indicator": "Orange"}
    ]
    
    # Pie Chart Data
    chart = {
        "title": _("Collection Breakdown (Export vs Domestic)"),
        "data": {
            "labels": [_("Export"), _("Domestic")],
            "datasets": [
                {
                    "name": _("Collection"),
                    "values": [export_collection, domestic_collection]
                }
            ]
        },
        "type": "donut",
        "colors": ["#10b981", "#f59e0b"]
    }
    
    return {
        "summary": summary,
        "chart": chart,
        "results": final_data
    }

@frappe.whitelist()
def export_to_excel(filters=None, export_type="all"):
    dashboard_data = get_dashboard_data(filters)
    data = dashboard_data.get("results")
    summary = dashboard_data.get("summary")
    
    if not data:
        return None

    wb = openpyxl.Workbook()
    
    # Based on export_type, we decide which sheets to keep
    if export_type == "all":
        ws_overview = wb.active
        ws_overview.title = "Dashboard Overview"
        ws_list = wb.create_sheet("Collection List")
    elif export_type == "detail":
        ws_list = wb.active
        ws_list.title = "Collection List"
        ws_overview = wb.create_sheet("Dummy") # Will be removed
    
    # Styling
    header_fill = PatternFill(start_color="2c3e50", fill_type="solid")
    header_font = Font(bold=True, color="FFFFFF")
    title_font = Font(bold=True, size=14)
    section_font = Font(bold=True, size=12)
    thin_side = Side(style='thin')
    table_border = Border(left=thin_side, right=thin_side, top=thin_side, bottom=thin_side)
    
    if export_type == "all":
        # 1. Overview Sheet
        ws_overview.cell(row=1, column=1, value="Collection Report Dashboard (Million INR)").font = title_font
        ws_overview.cell(row=1, column=4, value="Generated On: " + now_datetime().strftime("%Y-%m-%d %H:%M"))
        
        ws_overview.cell(row=3, column=1, value="Collection Metrics Summary").font = section_font
        
        colors = {"blue": "3b82f6", "green": "10b981", "orange": "f59e0b"}
        
        for i, s in enumerate(summary):
            r = 5 + (i // 3) * 3
            c = 1 + (i % 3) * 2
            bg_color = colors.get(s.get('indicator', 'blue').lower(), "3b82f6")
            
            cell_l = ws_overview.cell(row=r, column=c, value=s.get('label'))
            cell_l.font = Font(bold=True, color="FFFFFF")
            cell_l.fill = PatternFill(start_color=bg_color, fill_type="solid")
            cell_l.alignment = Alignment(horizontal="center")
            ws_overview.merge_cells(start_row=r, start_column=c, end_row=r, end_column=c+1)
            
            cell_v = ws_overview.cell(row=r+1, column=c, value=flt(s.get('value')) / 1000000)
            cell_v.font = Font(bold=True, size=11)
            cell_v.number_format = '"₹ "#,##0.0000" M"'
            cell_v.alignment = Alignment(horizontal="center")
            cell_v.border = Border(bottom=Side(style='medium', color=bg_color))
            ws_overview.merge_cells(start_row=r+1, start_column=c, end_row=r+1, end_column=c+1)

    if export_type in ["all", "detail"]:
        # 2. Collection List Sheet
        row_idx = 1
        ws_list.cell(row=row_idx, column=1, value="Detailed Collection List").font = section_font
        row_idx += 2
        
        headers = ["S.No.", "Invoice ID", "Date", "Customer", "Item", "Sales Person", "Amount (M)", "Type", "Status"]
        for idx, h in enumerate(headers, start=1):
            cell = ws_list.cell(row=row_idx, column=idx, value=h)
            cell.font, cell.fill, cell.alignment, cell.border = header_font, header_fill, Alignment(horizontal="center"), table_border
        row_idx += 1
        
        for r_idx, row in enumerate(data):
            ws_list.cell(row=row_idx, column=1, value=r_idx + 1).border = table_border
            ws_list.cell(row=row_idx, column=2, value=row['name']).border = table_border
            ws_list.cell(row=row_idx, column=3, value=row['posting_date']).border = table_border
            ws_list.cell(row=row_idx, column=4, value=row['customer']).border = table_border
            ws_list.cell(row=row_idx, column=5, value=row['item']).border = table_border
            ws_list.cell(row=row_idx, column=6, value=row['sales_person']).border = table_border
            
            amt_cell = ws_list.cell(row=row_idx, column=7, value=flt(row['allocated_amount']) / 1000000)
            amt_cell.number_format, amt_cell.border = '"₹ "#,##0.0000" M"', table_border
            
            ws_list.cell(row=row_idx, column=8, value="Export" if row['is_export'] else "Domestic").border = table_border
            ws_list.cell(row=row_idx, column=9, value=row['status']).border = table_border
            row_idx += 1
    
        # Add Total Row
        ws_list.cell(row=row_idx, column=1, value="Total").font = header_font
        ws_list.merge_cells(start_row=row_idx, start_column=1, end_row=row_idx, end_column=6)
        for c in range(1, 7):
            ws_list.cell(row=row_idx, column=c).fill = header_fill
            ws_list.cell(row=row_idx, column=c).border = table_border
            if c == 1:
                ws_list.cell(row=row_idx, column=c).alignment = Alignment(horizontal="right")
                
        total_amt = sum(flt(r['allocated_amount']) for r in data) / 1000000
        total_cell = ws_list.cell(row=row_idx, column=7, value=total_amt)
        total_cell.font = header_font
        total_cell.fill = header_fill
        total_cell.number_format, total_cell.border = '"₹ "#,##0.0000" M"', table_border
        
        ws_list.cell(row=row_idx, column=8, value="").fill = header_fill
        ws_list.cell(row=row_idx, column=8, value="").border = table_border
        ws_list.cell(row=row_idx, column=9, value="").fill = header_fill
        ws_list.cell(row=row_idx, column=9, value="").border = table_border
        row_idx += 1

    # Remove dummy if detail only
    if export_type == "detail" and "Dummy" in wb.sheetnames:
        wb.remove(wb["Dummy"])

    # Column Widths
    for ws in wb.worksheets:
        for i in range(1, 10):
            ws.column_dimensions[get_column_letter(i)].width = 20

    output = BytesIO()
    wb.save(output)
    output.seek(0)
    
    filename = f"Collection_Report_{nowdate()}.xlsx"
    if export_type == "detail":
        filename = f"Detailed_Collection_List_{nowdate()}.xlsx"
    
    return {
        "filename": filename,
        "filecontent": base64.b64encode(output.read()).decode()
    }

File: renu_customization/renu_customization/page/collection_report/collection_report.js
/* Collection Report Dashboard - Version 3.4 */
frappe.pages["collection_report"].on_page_load = function (wrapper) {
	console.log("Collection Report Dashboard - Version 3.4 (Stable Load)");
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Collection Report"),
		single_column: true,
	});

	// --- 1. SETUP CONTAINERS AND CSS ---
	$(`<style>
        @import url('https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700;800&display=swap');
        .layout-main-section { background-color: transparent !important; }
        .page-container { background-color: transparent !important; }
        .dashboard-content { padding: 20px; background: transparent !important; min-height: 100vh; font-family: 'Inter', sans-serif; color: #1e293b; width: 100% !important; }
        
        /* Filter Area Styling */
        .dashboard-filter-area { background: #fff; padding: 16px 24px; width: 100%; border-bottom: 1px solid #f1f5f9; }
        .dashboard-filter-area .section-body { display: flex !important; flex-wrap: wrap !important; gap: 12px !important; align-items: flex-end; }
        .dashboard-filter-area .frappe-control { margin-bottom: 0 !important; flex: 1; min-width: 180px; max-width: 240px; }
        .dashboard-filter-area .section-head { display: none !important; }

        /* Dashboard Cards */
        .summary-wrapper { 
            display: grid !important; 
            grid-template-columns: repeat(4, 1fr) !important; 
            gap: 16px; 
            margin-bottom: 24px; 
            width: 100% !important; 
        }
        .summary-card { background: #ffffff; border: 1px solid #e2e8f0; border-radius: 12px; padding: 16px; box-shadow: 0 1px 2px rgba(0, 0, 0, 0.05); border-left: 5px solid #cbd5e1; transition: all 0.3s ease; position: relative; }
        .summary-card:hover { transform: translateY(-4px); box-shadow: 0 12px 20px -5px rgba(0, 0, 0, 0.1); }
        .summary-card.blue { border-left-color: #3b82f6; }
        .summary-card.green { border-left-color: #10b981; }
        .summary-card.orange { border-left-color: #f59e0b; }
        .summary-card.purple { border-left-color: #8b5cf6; }

        .summary-card .label { font-size: 11px; color: #64748b; font-weight: 700; text-transform: uppercase; letter-spacing: 0.05em; margin-bottom: 8px; display: flex; align-items: center; gap: 8px; }
        .summary-card .value { font-size: 20px; font-weight: 800; color: #0f172a; }
        .summary-card .indicator { display: inline-block; width: 8px; height: 8px; border-radius: 50%; }
        .bg-blue { background-color: #3b82f6; }
        .bg-green { background-color: #10b981; }
        .bg-orange { background-color: #f59e0b; }
        .bg-purple { background-color: #8b5cf6; }

        /* Charts & Tables */
        .charts-row { display: grid; grid-template-columns: 1fr; gap: 24px; margin-bottom: 24px; width: 100%; }
        .chart-card { background: #fff; border-radius: 12px; padding: 24px; box-shadow: 0 1px 3px rgba(0,0,0,0.1); border: 1px solid #e2e8f0; }
        .chart-card .title { font-size: 13px; font-weight: 700; color: #1e293b; margin-bottom: 20px; text-transform: uppercase; letter-spacing: 0.05em; }
        .custom-legend { display: grid; grid-template-columns: repeat(auto-fill, minmax(200px, 1fr)); gap: 16px; margin-top: 30px; padding: 20px; border-top: 1px solid #f1f5f9; background: #fafafa; border-radius: 8px; }
        .legend-item { display: flex; align-items: flex-start; gap: 12px; }
        .legend-item .dot { width: 10px; height: 10px; border-radius: 2px; flex-shrink: 0; margin-top: 2px; }
        .legend-item .label { font-size: 11px; font-weight: 600; color: #475569; }
        .legend-item .val-pct { font-size: 10px !important; color: #94a3b8 !important; font-weight: 500 !important; }

        .table-card { background: #fff; border-radius: 12px; box-shadow: 0 1px 3px rgba(0,0,0,0.1); margin-bottom: 24px; overflow: hidden; border: 1px solid #e2e8f0; width: 100%; }
        .table-card .header { padding: 15px 24px; background: #fff; border-bottom: 1px solid #f1f5f9; font-weight: 700; color: #0f172a; display: flex; justify-content: space-between; align-items: center; }
        .dashboard-table { width: 100%; border-collapse: separate; border-spacing: 0; }
        .dashboard-table th { background: #f8fafc; padding: 12px 16px; text-align: left; font-size: 11px; font-weight: 700; color: #64748b; position: sticky; top: 0; z-index: 10; border-bottom: 1px solid #e2e8f0; text-transform: uppercase; }
        .dashboard-table td { padding: 12px 16px; border-bottom: 1px solid #f1f5f9; font-size: 13px; color: #334155; }
        .dashboard-table tr:hover td { background: #f8fafc; }
        
        tr.sticky-total td { position: sticky; bottom: 0; z-index: 30; background: #f8fafc !important; font-weight: 700; border-top: 2px solid #e2e8f0 !important; color: #0f172a; }

        .export-btn { font-size: 12px; cursor: pointer; color: #475569; font-weight: 600; padding: 6px 14px; border-radius: 6px; transition: all 0.2s; display: inline-flex; align-items: center; gap: 6px; background: #fff; border: 1px solid #e2e8f0; }
        .export-btn:hover { color: #2563eb !important; background: #eff6ff !important; border-color: #bfdbfe !important; }
        
        /* Indicators */
        .indicator-pill { padding: 2px 8px; border-radius: 12px; font-size: 10px; font-weight: 600; text-transform: uppercase; }
        .indicator-pill.Export { background: #ecfdf5; color: #065f46; }
        .indicator-pill.Domestic { background: #fff7ed; color: #9a3412; }

    </style>`).appendTo(page.main);

	let filter_parent = $('<div class="dashboard-filter-area"></div>').prependTo(page.main);
	page.container = $('<div class="dashboard-content"></div>').appendTo(page.main);

	// --- 2. DEFINE LOGIC FUNCTIONS ---
	function format_currency_short(num) {
		if (!num && num !== 0) return "₹ 0.0000 M";
		let value = flt(num) / 1000000;
		return (
			"₹ " +
			value.toLocaleString("en-US", { minimumFractionDigits: 4, maximumFractionDigits: 4 }) +
			" M"
		);
	}

	page.refresh = function () {
		let filters = page.filter_group ? page.filter_group.get_values() : {};

		// Show loading indicator
		if (page.container.is(":empty") || page.container.find(".summary-wrapper").length === 0) {
			page.container.html(
				'<div class="text-center" style="padding: 100px 0;"><i class="fa fa-refresh fa-spin fa-2x text-muted"></i><div class="mt-2 text-muted">Loading Collection Data...</div></div>',
			);
		} else {
			page.container.css("opacity", 0.6);
		}

		frappe.call({
			method: "renu_customization.renu_customization.page.collection_report.collection_report.get_dashboard_data",
			args: { filters: filters },
			callback: function (r) {
				page.container.css("opacity", 1);
				if (r.message) {
					page.dashboard_data = r.message;
					render_dashboard(r.message);
				}
			},
		});
	};

	function render_dashboard(data) {
		page.container.empty();
		if (!data.results || data.results.length === 0) {
			$(
				`<div class="text-center text-muted" style="padding: 100px 0;"><div>${__("No data found for selected criteria")}</div></div>`,
			).appendTo(page.container);
			return;
		}

		// KPI Cards
		let summary_row = $('<div class="summary-wrapper"></div>').appendTo(page.container);
		data.summary.forEach((metric) => {
			let indicator = (metric.indicator || "blue").toLowerCase();
			$(`
                <div class="summary-card ${indicator}">
                    <div class="label"><span class="indicator bg-${indicator}"></span>${metric.label}</div>
                    <div class="value">${format_currency_short(metric.value)}</div>
                </div>
            `).appendTo(summary_row);
		});

		// Chart Section
		let charts_row = $('<div class="charts-row"></div>').appendTo(page.container);
		let chart_card = $(`
            <div class="chart-card">
                <div class="title">${data.chart.title}</div>
                <div id="chart_wrapper" style="height: 350px;"></div>
                <div id="chart_legend" class="custom-legend"></div>
            </div>
        `).appendTo(charts_row);

		setTimeout(() => {
			page.chart = new frappe.Chart("#chart_wrapper", {
				data: data.chart.data,
				type: "donut",
				height: 350,
				colors: data.chart.colors,
				legend: 0,
				tooltipOptions: { formatTooltipY: (d) => format_currency_short(d) },
			});

			let legend_container = chart_card.find("#chart_legend");
			let total_val = data.chart.data.datasets[0].values.reduce((a, b) => a + b, 0);

			data.chart.data.labels.forEach((label, idx) => {
				let val = data.chart.data.datasets[0].values[idx];
				let color = data.chart.colors[idx % data.chart.colors.length];
				let share = total_val > 0 ? ((val / total_val) * 100).toFixed(1) + "%" : "0%";
				legend_container.append(`
                    <div class="legend-item">
                        <span class="dot" style="background: ${color}"></span>
                        <div class="info">
                            <span class="label">${label}</span>
                            <span class="val-pct">${format_currency_short(val)} (${share})</span>
                        </div>
                    </div>
                `);
			});
		}, 50);

		// Detailed Table
		let table_card = $(`
            <div class="table-card">
                <div class="header">
                    <span>${__("Detailed Collection List")}</span>
                    <div class="export-btn" id="export_excel_btn">
                        <i class="fa fa-file-excel-o"></i> Export to Excel
                    </div>
                </div>
                <div style="overflow: auto; width: 100%; max-height: 800px;">
                    <table class="dashboard-table">
                        <thead>
                            <tr>
                                <th>${__("Payment ID")}</th>
                                <th>${__("Invoice ID")}</th>
                                <th>${__("Date")}</th>
                                <th>${__("Customer")}</th>
                                <th>${__("Item")}</th>
                                <th>${__("Sales Person")}</th>
                                <th style="text-align: right;">${__("Amount")}</th>
                                <th style="text-align: center;">${__("Type")}</th>
                            </tr>
                        </thead>
                        <tbody id="collection_table_body"></tbody>
                    </table>
                </div>
            </div>
        `).appendTo(page.container);

		let tbody = table_card.find("#collection_table_body");
		let total_amt = 0;
		data.results.forEach((row) => {
			let type_label = row.is_export ? "Export" : "Domestic";
			let display_val = Math.round((flt(row.allocated_amount) / 1000000) * 10000) / 10000;
			total_amt += display_val;
			$(`
                <tr>
                    <td><a href="/app/payment-entry/${row.payment_entry}" style="font-weight: 600; color: #4338ca;">${row.payment_entry}</a></td>
                    <td><a href="/app/sales-invoice/${row.name}" style="font-weight: 500; color: #64748b;">${row.name}</a></td>
                    <td>${frappe.datetime.str_to_user(row.posting_date)}</td>
                    <td>${row.customer}</td>
                    <td>
                        <div style="line-height: 1.4;">
                            <div style="font-size: 11px; color: #64748b;">${row.item_code || "-"}</div>
                            <div style="font-weight: 600;">${row.item_name || "-"}</div>
                        </div>
                    </td>
                    <td>${row.sales_person || "-"}</td>
                    <td style="text-align: right; font-weight: 700;">${format_currency_short(row.allocated_amount)}</td>
                    <td style="text-align: center;"><span class="indicator-pill ${type_label}">${__(type_label)}</span></td>
                </tr>
            `).appendTo(tbody);
		});

		// Grand Total Footer
		let total_display = "₹ " + total_amt.toLocaleString("en-US", { minimumFractionDigits: 4, maximumFractionDigits: 4 }) + " M";
		$(`
			<tfoot>
				<tr class="sticky-total">
					<td colspan="6" style="text-align: right; padding-right: 20px;">GRAND TOTAL</td>
					<td style="text-align: right; font-weight: 800; border-left: 1px solid #e2e8f0; background: #f8fafc;">${total_display}</td>
					<td></td>
				</tr>
			</tfoot>
		`).appendTo(table_card.find(".dashboard-table"));

		table_card.find("#export_excel_btn").click(() => export_to_excel("detail"));
	}

	// --- 3. INITIALIZE FILTERS AND ACTIONS ---
	const filter_fields = [
		{ fieldname: "fiscal_year", label: __("Fiscal Year"), fieldtype: "Link", options: "Fiscal Year", placeholder: __("Select Year") },
		{ fieldtype: "Column Break" },
		{ fieldname: "from_date", label: __("From Date"), fieldtype: "Date", placeholder: __("Start Date") },
		{ fieldtype: "Column Break" },
		{ fieldname: "to_date", label: __("To Date"), fieldtype: "Date", placeholder: __("End Date") },
		{ fieldtype: "Column Break" },
		{ fieldname: "customer", label: __("Customer"), fieldtype: "Link", options: "Customer", placeholder: __("Select Customer") },
		{ fieldtype: "Column Break" },
		{ fieldname: "sales_person", label: __("Sales Person"), fieldtype: "Link", options: "Sales Person", placeholder: __("Select Sales Person") },
		{ fieldtype: "Column Break" },
		{ fieldname: "dom_exp", label: __("Type"), fieldtype: "Select", options: ["", "Domestic", "Export"], placeholder: __("Select Type") },
	];

	page.filter_group = new frappe.ui.FieldGroup({ parent: filter_parent, fields: filter_fields });
	page.filter_group.make();

	// Attach events
	Object.keys(page.filter_group.fields_dict).forEach((key) => {
		let field = page.filter_group.fields_dict[key];
		field.on_change = () => page.refresh();
	});

	page.set_primary_action(__("Refresh"), () => page.refresh());
	page.add_menu_item(__("Export to PDF"), () => export_pdf());
	page.add_menu_item(__("Export to Excel"), () => export_to_excel("all"));

	// --- 4. EXPORT FUNCTIONS ---
	const export_to_excel = (export_type = "all") => {
		const filters = page.filter_group.get_values();
		frappe.show_alert({ message: __("Generating Excel Report..."), indicator: "blue" });
		frappe.call({
			method: "renu_customization.renu_customization.page.collection_report.collection_report.export_to_excel",
			args: { filters: filters, export_type: export_type },
			callback: function (r) {
				if (r.message) {
					const { filename, filecontent } = r.message;
					const byteCharacters = atob(filecontent);
					const byteNumbers = new Array(byteCharacters.length);
					for (let i = 0; i < byteCharacters.length; i++) byteNumbers[i] = byteCharacters.charCodeAt(i);
					const blob = new Blob([new Uint8Array(byteNumbers)], { type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" });
					const link = document.createElement("a");
					link.href = window.URL.createObjectURL(blob);
					link.download = filename;
					link.click();
					frappe.show_alert({ message: __("Excel Report Downloaded"), indicator: "green" });
				}
			},
		});
	};

	const export_pdf = async () => {
		if (!page.dashboard_data) return;
		frappe.show_alert({ message: __("Preparing PDF..."), indicator: "blue" });
		const get_chart_png = () => {
			const svg = document.querySelector(`#chart_wrapper svg`);
			if (!svg) return null;
			const canvas = document.createElement("canvas");
			const context = canvas.getContext("2d");
			const svg_data = new XMLSerializer().serializeToString(svg);
			const img = new Image();
			return new Promise((resolve) => {
				img.onload = () => {
					canvas.width = img.width * 2; canvas.height = img.height * 2;
					context.fillStyle = "white"; context.fillRect(0, 0, canvas.width, canvas.height);
					context.drawImage(img, 0, 0, canvas.width, canvas.height);
					resolve(canvas.toDataURL("image/png"));
				};
				img.src = "data:image/svg+xml;base64," + btoa(unescape(encodeURIComponent(svg_data)));
			});
		};

		const chart_png = await get_chart_png();
		const data = page.dashboard_data;
		const total_val = data.chart.data.datasets[0].values.reduce((a, b) => a + b, 0) || 1;
		const period = (page.filter_group.get_values().from_date) ? `${page.filter_group.get_values().from_date} to ${page.filter_group.get_values().to_date}` : "Current Selection";

		let html = `<html><head><style>
            body { font-family: sans-serif; font-size: 10px; }
            .header { text-align: center; border-bottom: 2px solid #3b82f6; padding-bottom: 10px; margin-bottom: 20px; }
            .kpi-wrapper { display: flex; gap: 10px; margin-bottom: 20px; }
            .kpi-card { flex: 1; border: 1px solid #eee; padding: 10px; text-align: center; background: #fafafa; border-radius: 5px; }
            table { width: 100%; border-collapse: collapse; margin-top: 20px; }
            th, td { border: 1px solid #eee; padding: 6px; text-align: left; }
            th { background: #f4f4f4; }
        </style></head><body>
            <div class="header"><h1>Collection Report</h1><p>${period}</p></div>
            <div class="kpi-wrapper">${data.summary.map(m => `<div class="kpi-card"><div>${m.label}</div><div style="font-size:14px;font-weight:bold;">${format_currency_short(m.value)}</div></div>`).join("")}</div>
            <div style="text-align:center;"><img src="${chart_png}" style="max-width:400px;"></div>
            <h3>Detailed Collection List</h3>
            <table><thead><tr><th>Payment ID</th><th>Invoice ID</th><th>Date</th><th>Customer</th><th>Amount</th><th>Type</th></tr></thead>
            <tbody>${data.results.map(row => `<tr><td>${row.payment_entry}</td><td>${row.name}</td><td>${row.posting_date}</td><td>${row.customer}</td><td>${format_currency_short(row.allocated_amount)}</td><td>${row.is_export ? "Export" : "Domestic"}</td></tr>`).join("")}</tbody>
            </table></body></html>`;

		const $form = $(`<form action="/api/method/renu_customization.renu_customization.page.collection_report.collection_report.export_to_pdf" method="POST" target="_blank" style="display:none;"><input type="hidden" name="html" value=""><input type="hidden" name="csrf_token" value="${frappe.csrf_token}"></form>`).appendTo("body");
		$form.find('input[name="html"]').val(html);
		$form.submit();
		$form.remove();
	};

	// --- 5. INITIAL LOAD ---
	page.refresh();
};
